### PR TITLE
TF-3887 Add SMTP Settings

### PR DESCRIPTION
**Pull request by chpag**
## Description

This PR adds support for managing Admin SMTP settings in Terraform Enterprise by introducing a new resource and data source for Admin SMTP configuration.

- [X] _Update the [Change Log](https://github.com/hashicorp/terraform-provider-tfe/blob/main/docs/changelog-process.md)_
- [X] _Update the [Documentation](https://github.com/hashicorp/terraform-provider-tfe/blob/main/docs/changelog-process.md#updating-the-documentation)_

## Testing plan

1. Configure Terraform Enterprise with admin token
2. Create SMTP settings resource with no authentication:
```terraform
provider "tfe" {
  hostname = "tfe.example.com"
  token    = var.admin_token
}

resource "tfe_admin_smtp_settings" "test" {
  host   = "smtp.example.com"
  port   = 25
  sender = "noreply@example.com"
  auth   = "none"
}
```
3. Apply and verify SMTP settings are created
4. Update to use authentication with write-only password:
```terraform
resource "tfe_admin_smtp_settings" "test" {
  host                = "smtp.example.com"
  port                = 587
  sender              = "noreply@example.com"
  auth                = "login"
  username            = "smtp_user"
  password_wo         = var.smtp_password
  password_wo_version = 1
}
```
5. Apply and verify settings are updated
6. Test data source to read SMTP settings:
```terraform
data "tfe_admin_smtp_settings" "current" {}

output "smtp_enabled" {
  value = data.tfe_admin_smtp_settings.current.enabled
}
```
7. Verify import functionality: terraform import tfe_admin_smtp_settings.test smtp
8. Test destroy operation (should disable SMTP, not delete)

## External links

[go-tfe SMTP Settings API](https://pkg.go.dev/github.com/hashicorp/go-tfe#AdminSMTPSetting)
[TFE Admin Settings API documentation](https://developer.hashicorp.com/terraform/enterprise/api-docs/admin/settings)

## Output from acceptance tests

_Please run applicable acceptance tests locally and include the output here. See [testing.md](https://github.com/hashicorp/terraform-provider-tfe/blob/main/docs/testing.md) to learn how to run acceptance tests._

_If you are an external contributor, your contribution(s) will first be reviewed before running them against the project's CI pipeline._

```
$ TF_ACC=1 ENABLE_TFE=1 go test -v ./internal/provider -run TestAccTFEAdminSMTPSettings -timeout 30m -count=1

=== RUN   TestAccTFEAdminSMTPSettingsDataSource_basic
--- PASS: TestAccTFEAdminSMTPSettingsDataSource_basic (1.38s)
=== RUN   TestAccTFEAdminSMTPSettings_omnibus
=== RUN   TestAccTFEAdminSMTPSettings_omnibus/basic_SMTP_settings_without_authentication
--- PASS: TestAccTFEAdminSMTPSettings_omnibus (1.13s)
    --- PASS: TestAccTFEAdminSMTPSettings_omnibus/basic_SMTP_settings_without_authentication (1.13s)
=== RUN   TestAccTFEAdminSMTPSettings_AuthNone
--- PASS: TestAccTFEAdminSMTPSettings_AuthNone (1.08s)
=== RUN   TestAccTFEAdminSMTPSettings_AuthPlain_writeOnly
--- PASS: TestAccTFEAdminSMTPSettings_AuthPlain_writeOnly (1.12s)
=== RUN   TestAccTFEAdminSMTPSettings_AuthLogin_writeOnly
--- PASS: TestAccTFEAdminSMTPSettings_AuthLogin_writeOnly (1.07s)
=== RUN   TestAccTFEAdminSMTPSettings_AuthPlain_writeOnly_update
--- PASS: TestAccTFEAdminSMTPSettings_AuthPlain_writeOnly_update (1.89s)
=== RUN   TestAccTFEAdminSMTPSettings_AuthPlain_writeOnly_no_version_change
--- PASS: TestAccTFEAdminSMTPSettings_AuthPlain_writeOnly_no_version_change (1.71s)
=== RUN   TestAccTFEAdminSMTPSettings_AuthPlain
--- PASS: TestAccTFEAdminSMTPSettings_AuthPlain (1.08s)
=== RUN   TestAccTFEAdminSMTPSettings_AuthLogin
--- PASS: TestAccTFEAdminSMTPSettings_AuthLogin (1.09s)
=== RUN   TestAccTFEAdminSMTPSettings_TestEmailAddressDrift
--- PASS: TestAccTFEAdminSMTPSettings_TestEmailAddressDrift (1.72s)
=== RUN   TestAccTFEAdminSMTPSettings_HostSenderNullPreservation
--- PASS: TestAccTFEAdminSMTPSettings_HostSenderNullPreservation (1.81s)
PASS
ok  	github.com/hashicorp/terraform-provider-tfe/internal/provider	15.815s

<!-- heimdall_github_prtemplate:grc-pci_dss-2024-01-05 -->
## Rollback Plan

<!--
Please outline a plan in the event changes need to be rolled back

Example: If a change needs to be reverted, we will roll out an update to the code within 7 days.
-->

## Changes to Security Controls

<!--
Are there any changes to security controls (access controls, encryption, logging) in this pull request? If so, explain.
-->

### CODE REVIEW

**Review by ctrombley — CHANGES_REQUESTED** [review]
This is looking really good, thanks @chpag. I found a couple edge cases where it looks like perpetual config drift can be introduced. 

1. `modelFromTFEAdminSMTPSettings` unconditionally writes `types.StringValue(v.Host)` and `types.StringValue(v.Sender)`. If the user removes these from config, the update sends "" to the API, but Read then writes "" back to state. Terraform sees null in the config vs "" in state and drift occurs.
    -  Fix: `resource_tfe_smtp_settings.go ~L200-210` — Extend null value preservation to `Host` and `Sender` in the `Read`, `Create`. and `Update` methods:
```go
if m.Host.IsNull() {
    result.Host = types.StringNull()
}
if m.Sender.IsNull() {
    result.Sender = types.StringNull()
}
```
This can be extracted into a single method to cut down on the code repetition between the methods.


2. `test_email_address` is never returned by the API, so `Read` always sets it to null in state. This can lead to perpetual drift.
  - Fix: preserve the field from prior state/config in `Read`, `Create`, and `Update`, the same way `password` and `password_wo_version` are already handled:
```go
  // test_email_address is never returned by the API — always preserve from config
  result.TestEmailAddress = config.TestEmailAddress

```
`UseStateForUnknown` won't work here — it only fires when the planned value is unknown, but after `Read` the value becomes null (known), so the modifier never triggers.

Here's a [gist with some tests](https://gist.github.com/ctrombley/7941cea88561efa2b3a19bea3fa15e67) that expose the issues, feel free to use those in your PR if desired. 

(I'm not sure if you have an actual TFE instance to run tests against, but I was able to get these running on my local instance in cloud mode by replacing the implementation of `BasePunditPolicy#admin_api_access_limited` to always return false.)

**Review by brandonc — COMMENTED** [review]
Sorry to come in late but I'm wondering if you will consider renaming this resource to be `tfe_admin_smtp_settings` so that it appears in documentation next to `tfe_admin_organization_settings` ?

**Comment by chpag** [comment]
@ctrombley :  I have update the code regarding:

-  the Preservation of Null for Sender and Host, and put all this Preservation Code in a single method `preserveOptionalFields`, to avoid code repetition
- Add test_email_address to preserveOptionFields, to fix the fact that it's not return by API
- Add your test example, regarding  Drift, and null preservation
- Rename resource/datasource from tfe_smtp_settings to tfe_admin_smtp_settings in code ans also documentation.
- Relaunch The Test on TFE and update the PR Description with latest test result

**Review by Uk1288 — DISMISSED** [review]
LGTM! Ran the tfe nightly against this branch and its all ✅ passing as expected.

File: internal/provider/resource_tfe_admin_smtp_settings.go [hunk 1]
```diff
@@ -0,0 +1,405 @@
...
+				},
+			},
+			"password_wo": schema.StringAttribute{
+				Description: "The password in write only used to authenticate to the SMTP server. Required if auth is 'login' or 'plain'.",
+				Optional:    true,
+				Sensitive:   true,
+				WriteOnly:   true,
+				Validators: []validator.String{
+					stringvalidator.ConflictsWith(path.MatchRoot("password")),
+					validators.AttributeValueConflictValidator("auth", []string{string(tfe.SMTPAuthNone)}),
```
Comment (ctrombley) [inline internal/provider/resource_tfe_admin_smtp_settings.go:159]: Can we add `AlsoRequires(path.MatchRoot("password_wo_version")` here since both attributes require the other?

**Review by ctrombley — APPROVED** [review]
Looks good, thanks @chpag !
